src/model/trader.rs

```diff
@@ -212,33 +224,79 @@ pub fn settle_trade_on_counterparty(mut counterparty: Trader, offer: &Offer) ->
     //     // Do nothing.
     //     return counterparty
     // }
-
-    println!(
-        "***** TRADE SETTLED FOR {:?} WITH TRADER {} *****",
-        offer,
-        counterparty.id()
-    );
-    println!(
-        "Prior inventory: Food: {}, Water: {}",
-        counterparty.forager.count(&Resource::Food),
-        counterparty.forager.count(&Resource::Water)
-    );
+    if core_config().simulation.VERBOSITY > 1 {
+        println!(
+            "***** TRADE SETTLED FOR {:?} WITH TRADER {} *****",
+            offer,
+            counterparty.id()
+        );
+        println!(
+            "Prior inventory: Food: {}, Water: {}",
+            counterparty.forager.count(&Resource::Food),
+            counterparty.forager.count(&Resource::Water)
+        );
+    }
     // Settle inventories.
     counterparty.acquire(&Resource::Food, -1 * offer.food_delta());
     counterparty.acquire(&Resource::Water, -1 * offer.water_delta());
-    println!(
-        "Final inventory: Food: {}, Water: {}",
-        counterparty.forager.count(&Resource::Food),
-        counterparty.forager.count(&Resource::Water)
-    );
-
+    if core_config().simulation.VERBOSITY > 1 {
+        println!(
+            "Final inventory: Food: {}, Water: {}",
+            counterparty.forager.count(&Resource::Food),
+            counterparty.forager.count(&Resource::Water)
+        );
+    }
     counterparty
 }
 
 impl Agent for Trader {
     fn step(&mut self, state: &mut dyn krabmaga::engine::state::State) {
-        // all trades were facilitated in `before_step` function on the `board`
-        // next the agents choose an action given the agent_state post trading
+        let board = state.as_any_mut().downcast_mut::<Board>().unwrap();
+        if (board.step > 0) & board.has_trading {
+            // Traders state is fixed so can be used to retrieve trader
+            let mut traders = board.get_agents();
+            traders.shuffle(&mut board.rng);
+
+            // Execute trade if available.
+            if !self.offer().is_trivial() {
+                if !board.traded.contains_key(&self.id()) {
+                    // println!("Not traded yet");
+                    let offer = self.offer();
+                    for counterparty in &traders {
+                        // println!("Counterparty: {}", counterparty);
+                        let counterparty_id = counterparty.id();
+                        // If already traded, continue
+                        if board.traded.contains_key(&counterparty_id) {
+                            continue;
+                        }
+                        if counterparty_id != self.id()
```

Comment:
don't think we need `counterparty_id != self.id()` because we have already checked that `self.id` is not in `traded`

Reply:
I've added a comment to clarify - I think we still need this in general as this checks the counterparty is not the same as self to avoid self trading with self. However, our current offer matching means self should not match with itself so it is currently redundant but I think worth keeping for clarity.

Reply:
Agreed, sorry I got in a muddle